File: 15-reverse-linked-list/solution.go
// Reverse Linked List (Iterative and Recursive)
// Time: O(n), Space: O(1) iterative / O(n) recursive

package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

type Solution struct{}

func (s *Solution) ReverseListIterative(head *ListNode) *ListNode {
	var prev *ListNode
	curr := head
	
	for curr != nil {
		nextTemp := curr.Next
		curr.Next = prev
		prev = curr
		curr = nextTemp
	}
	
	return prev
}

func (s *Solution) ReverseListRecursive(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	
	newHead := s.ReverseListRecursive(head.Next)
	head.Next.Next = head
	head.Next = nil
	
	return newHead
}

func main() {
	head := &ListNode{Val: 1, Next: &ListNode{Val: 2, Next: &ListNode{Val: 3}}}
	sol := &Solution{}
	reversedHead := sol.ReverseListIterative(head)
	curr := reversedHead
	for curr != nil {
		if curr.Next != nil {
			fmt.Printf("%d -> ", curr.Val)
		} else {
			fmt.Printf("%d\n", curr.Val)
		}
		curr = curr.Next
	}
}

File: 15-reverse-linked-list/solution.js
// Reverse Linked List (Iterative and Recursive)
// Time: O(n), Space: O(1) iterative / O(n) recursive

class ListNode {
    constructor(val = 0, next = null) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    reverseListIterative(head) {
        let prev = null;
        let curr = head;
        
        while (curr) {
            const nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }
        
        return prev;
    }
    
    reverseListRecursive(head) {
        if (!head || !head.next) {
            return head;
        }
        
        const newHead = this.reverseListRecursive(head.next);
        head.next.next = head;
        head.next = null;
        
        return newHead;
    }
}

// Test
if (require.main === module) {
    const head = new ListNode(1, new ListNode(2, new ListNode(3)));
    const sol = new Solution();
    let reversedHead = sol.reverseListIterative(head);
    let curr = reversedHead;
    const result = [];
    while (curr) {
        result.push(curr.val);
        curr = curr.next;
    }
    console.log(result.join(" -> "));  // 3 -> 2 -> 1
}

module.exports = { Solution, ListNode };
